### PR TITLE
Iox #482 Add example overview to website

### LINKED ISSUE #482: Extend documentation for 1.0 release and website
# Documentation epic

# Proposed outline
- index.md
- What is Eclipse iceoryx? (Explain the three pillars easy, safe, fast)
- getting-started
  - overview.md (Split?)
    - General intro (add overview pictures e.g. from [ROSCon slides](https://roscon.ros.org/2019/talks/roscon2019_truezerocopy.pdf), add section "limitations")
    - hello world/quick start: icedelivery c++ typed (without `and_then`?) `icehello`
    - User API (Move into examples)
      - Typed
      - Untyped  
  - installation.md (move colcon to advanced?, usual user just wants to use script and nothing more, link contributor installation guide, move iceoryx_posh options table to advanced, fix markdown no identation before code blocks)
  - examples (overview over examples)
    - Communication with polling for the user /for frameworks
    - Options
    - Listner 
    - Communication without polling
    - Debugging (iceperf, icecrystal)
- API-reference
- (API-cheatsheet.md)
- FAQ.md
- advanced
  - Contriubtor guides 
    - best-practice-for-testing.md
    - Modern C++ error handling (Optional, expected)
    - installation-guide-for-contributors.md 
  - iceoryx_utils.md
  - usage-guide.md

## Documents covering
- [x] Extend installation guide
  - [x] QNX7 with QCC 5.4.0
- [x] Introduction: motivation, goals and capabilities of iceoryx
  - [x] Which data types are allowed to be sent via iceoryx? What are the limitations (virtual, vtable, heap)?
  - [x] Clarify failure cases, e.g.
    - What happens if an application segfaults? If crashes in critical section memleak
    - Does the crash of one application affect the whole system
    - When an application crashes, what happens with its resources like memory chunks, UDS, etc.
  - [x] Update Waitset and Listener section in overview.md and have a use case based comparison 
- [x]  Add new examples
  - [x] README.md in the example 
  - [x] iceoptions 
  - [x] icehello
  - [x] iceensemble: Explain how C and C++ API work together (multiple publishers and subscribers)
  - [x] blocking publisher
  - [x] icecubetray: Add access control documentation (comes from #640)
  - [x] Record asciinema for all examples
- [x] Add example for static RouDi Config with custom mempools and one for the dynamic TOML config, e.b. by just a 
- [x] `@todo` in the code shall not be linked to issues for v1.0
- [x] Add Apex.OS to the framework section

### v2.0 Release (Moved to #743)

- [x] Improve utils documentation (in the long run: find more suitable name for utils, e.g. safe building blocks), see also #468
- [x] Describe all RouDi command line options?
- [x] Add info about release names (ice cream flavours)
- [x] add overview graphic introducing the main players (RouDi, Runtime/user apps, shared Memory, data packages/topics) in overview.md (see suggestions [here](https://github.com/eclipse-iceoryx/iceoryx/pull/646#discussion_r604950011) and [here](https://github.com/eclipse-iceoryx/iceoryx/pull/646#discussion_r604960591))
- [x] Add chunk header example
- [x] Add example sending complex data types
- [x] icedelivery:  Explain lifetime semantics of chunks/samples and user restrictions: loan needs to be followed with publish or release, MAX_CHUNKS_IN_USE on publisher and subscriber side etc., API is not thread-safe

Consolidate current documentation as far as possible (keep the scope of the issue confined to documentation).
Note: doxygen documentation rework is out of scope here and belongs to #468

This is later to be integrated into the website #136

### PR DESCRIPTION
**Pull request by mossmaurice**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md

## Notes for Reviewer
* Restructure examples for v1.0 Almond release
* No example docu/code touched yet
* Add `mike` dependency for versioning MkDocs

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Partly closes #482

### CODE REVIEW

File: doc/website/getting-started/examples/iceoptions.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+title: Configure the history cache size and the starting behaviour
```
Comment (dkroenke) [inline doc/website/getting-started/examples/iceoptions.md:2]: Configure the history cache size and startup behavior

File: iceoryx_examples/README.md [hunk 2]
```diff
@@ -1,42 +1,3 @@
...
-
-## How to add a new example
-
- 1. Add the example in the "List of all examples" in this markdown file
- 2. Add the example folder name into the `EXAMPLES=${EXAMPLES} ...` array in `./tools/iceoryx_build_test.sh`
- 3. Add an `add_subdirectory` directive into `iceoryx_meta/CMakeLists.txt` in the `if(EXAMPLES)` section.
- 4. Add the example in the file with the corresponding level in `doc/website/getting-started/examples/LEVEL-examples.md`
+# Examples for iceoryx
 
+Please visit [iceoryx.io/getting-started/list-of-examples](http://iceoryx.io/getting-started/list-of-examples/) to find a description of the examples.
```
Comment (dkroenke) [inline iceoryx_examples/README.md:3]: maybe use https in the link?

File: doc/website/getting-started/list-of-examples.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# List of examples
+
+| Example                                         | Description                                                 | Level              |
+|:------------------------------------------------|:------------------------------------------------------------|:-------------------|
+|[icehello](./examples/icehello/)                 | How to send data to another process                         | :star:             |
+|[icedelivery](./examples/icedelivery/)           | How to send and receive data in a functional way            | :star:             |
+|[iceoptions](./examples/iceoptions/)             | Configure the history cache size and the starting behaviour | :star:             |
```
Comment (dkroenke) [inline doc/website/getting-started/list-of-examples.md:7]: It is hard to understand what is meant with "starting behaviour".
Maybe this can be more generic like:
"Extended Configuration of behavior for Publisher/Subscriber Ports"

File: doc/website/getting-started/list-of-examples.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+|[callbacks](./examples/callbacks/)               | Implement callbacks which are triggered by events           | :star::star:       |
+|[callbacks_in_c](./examples/callbacks_in_c/)     | How to implement callbacks in C                             | :star::star:       |
+|[waitset](./examples/waitset/)                   | Wait until one or more subscriber has received data         | :star::star:       |
+|[waitset_in_c](./examples/waitset_in_c/)         | How to use the WaitSet in C                                 | :star::star:       |
+|[iceensemble](./examples/iceensemble/)           | How to make use of the n:m communication feature            | :star::star:       |
+|[singleprocess](./examples/singleprocess/)       | How to communicate in a single process between threads      | :star::star:       |
+|[iceperf](./examples/iceperf/)                   | Measures the latency of different IPC mechanisms            | :star::star::star: |
+|[icecrystal](./examples/icecrystal/)             | How to use the introspection client                         | :star::star::star: |
```
Comment (dkroenke) [inline doc/website/getting-started/list-of-examples.md:16]: How to use the introspection client for debugging the middleware

File: doc/website/getting-started/list-of-examples.md (hunk 3 above)
Comment (mossmaurice) [reply doc/website/getting-started/list-of-examples.md:7]: I'd explicitly wanted to mention an example or the point that is most important from user perspective. In terms of issues on GitHub it was definitely the "I missed the first sample in the example" issue. I suppose "Extended configuration" wouldn't guide the user to this example when reading over the list, what do you think about "Configure the history cache size and the starting behaviour of publishers and subscribers"

File: doc/website/getting-started/list-of-examples.md (hunk 4 above)
Comment (mossmaurice) [reply doc/website/getting-started/list-of-examples.md:16]: I'll leave out "the middleware" as this does not add value when reading over the list. In the context of iceoryx this should be clear.

File: iceoryx_examples/README.md (hunk 2 above)
Comment (mossmaurice) [reply iceoryx_examples/README.md:3]: I double checked yesterday and think GitHub pages doesn't support https, what a shame :(

Comment (dkroenke) [reply iceoryx_examples/README.md:3]: oof ok

File: doc/website/getting-started/examples/icehello.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+title: How to send data to another process
```
Comment (elfenpiff) [inline doc/website/getting-started/examples/icehello.md:2]: This is also somehow confusing. Maybe just
Start communicating via iceoryx or how to send data to another process.

File: doc/website/getting-started/examples/iceperf.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+title: Measures the latency of different IPC mechanisms
```
Comment (elfenpiff) [inline doc/website/getting-started/examples/iceperf.md:2]: Measuring

File: doc/website/getting-started/examples/waitset.md [hunk 7]
```diff
@@ -0,0 +1,4 @@
+---
+title: Wait until one or more subscriber has received data
```
Comment (elfenpiff) [inline doc/website/getting-started/examples/waitset.md:2]: Getting notified by events like the receiving of data (with the WaitSet).

File: doc/website/getting-started/examples/waitset_in_c.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+---
+title: How to use the WaitSet in C
```
Comment (elfenpiff) [inline doc/website/getting-started/examples/waitset_in_c.md:2]: Getting notified by events like the receiving of data (with the WaitSet) by using the C API.

File: doc/website/getting-started/list-of-examples.md [hunk 9]
```diff
@@ -0,0 +1,16 @@
+# List of examples
+
+| Example                                         | Description                                                 | Level              |
+|:------------------------------------------------|:------------------------------------------------------------|:-------------------|
+|[icehello](./examples/icehello/)                 | How to send data to another process                         | :star:             |
```
Comment (elfenpiff) [inline doc/website/getting-started/list-of-examples.md:5]: This example is missing. Therefore I would wait with a merge of this document until the examples are available or we remove those examples which are not present for now and add them later.

File: doc/website/getting-started/list-of-examples.md [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+|[icedelivery](./examples/icedelivery/)           | How to send and receive data in a functional way            | :star:             |
```
Comment (elfenpiff) [inline doc/website/getting-started/list-of-examples.md:6]: We do not send or receive data in a functional way, only the error handling is somehow functional when it is used in the `.and_then` `.or_else` fashion.

File: doc/website/getting-started/list-of-examples.md [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+|[iceoptions](./examples/iceoptions/)             | Configure the history cache size and the starting behaviour | :star:             |
+|[icedelivery_in_c](./examples/icedelivery_in_c/) | How to send and receive data using the C API                | :star:             |
+|[callbacks](./examples/callbacks/)               | Implement callbacks which are triggered by events           | :star::star:       |
```
Comment (elfenpiff) [inline doc/website/getting-started/list-of-examples.md:9]: Implement event triggered callbacks.

File: doc/website/getting-started/list-of-examples.md [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+|[callbacks_in_c](./examples/callbacks_in_c/)     | How to implement callbacks in C                             | :star::star:       |
```
Comment (elfenpiff) [inline doc/website/getting-started/list-of-examples.md:10]: Implement event triggered callbacks with the C API

File: doc/website/getting-started/list-of-examples.md [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+|[waitset](./examples/waitset/)                   | Wait until one or more subscriber has received data         | :star::star:       |
```
Comment (elfenpiff) [inline doc/website/getting-started/list-of-examples.md:11]: Getting notified by events like the receiving of data (with the WaitSet).

File: doc/website/getting-started/list-of-examples.md [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+|[waitset_in_c](./examples/waitset_in_c/)         | How to use the WaitSet in C                                 | :star::star:       |
```
Comment (elfenpiff) [inline doc/website/getting-started/list-of-examples.md:12]: Getting notified by events like the receiving of data with the C API.

File: doc/website/getting-started/list-of-examples.md [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+|[iceensemble](./examples/iceensemble/)           | How to make use of the n:m communication feature            | :star::star:       |
```
Comment (elfenpiff) [inline doc/website/getting-started/list-of-examples.md:13]: Using multiple publisher for one topic (or how to use n:m communication).

File: doc/website/getting-started/list-of-examples.md [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+|[singleprocess](./examples/singleprocess/)       | How to communicate in a single process between threads      | :star::star:       |
+|[iceperf](./examples/iceperf/)                   | Measures the latency of different IPC mechanisms            | :star::star::star: |
```
Comment (elfenpiff) [inline doc/website/getting-started/list-of-examples.md:15]: Measuring

File: doc/website/getting-started/list-of-examples.md [hunk 17]
```diff
@@ -0,0 +1,16 @@
+# List of examples
+
+| Example                                         | Description                                                 | Level              |
```
Comment (elfenpiff) [inline doc/website/getting-started/list-of-examples.md:3]: Please order the list alphabetically.

File: iceoryx_examples/README.md (hunk 2 above)
Comment (elfenpiff) [reply iceoryx_examples/README.md:3]: Please be aware that most people use the github pages when looking at the documentation! And with this approach we have another indirection where you switch from github to website to github again ... I do not like this. 
Please take a look at this here: https://github.com/IntelRealSense/librealsense/tree/master/examples

Could we please move the list of examples back in here!

For the website I suggest that we should somehow adjust the website deployment script so that this file is moved/copied to the `website/getting-started/list-of-examples.md` and then handled by the generation script.

So in short.
1. This file contains all the contents of `list-of-examples.md`.
2. When calling mkdoc or whatever this file is copied to `website/getting-started/list-of-examples.md` temporarily and the website is generated
3. after generating the website the copied file is removed again

File: doc/website/getting-started/list-of-examples.md (hunk 17 above)
Comment (elfenpiff) [inline doc/website/getting-started/list-of-examples.md:3]: Could we add another row called technology where we state the technology which the example is presenting. Like for instance:
- WaitSet
- Listener
- n:m Communication
- IPC
- InterThreadCommunication
- Introspection

**Review by elfenpiff — CHANGES_REQUESTED** [review]

File: doc/website/getting-started/examples/callbacks.md [hunk 18]
```diff
@@ -0,0 +1,5 @@
+---
+title: Implement callbacks which are triggered by events
```
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply doc/website/getting-started/examples/callbacks.md:2]: I would omit  "(by using the Listener)" and only mention the "what" not the "how" here

File: doc/website/getting-started/examples/callbacks_in_c.md [hunk 19]
```diff
@@ -0,0 +1,5 @@
+---
+title: How to implement callbacks in C
```
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply doc/website/getting-started/examples/callbacks_in_c.md:2]: I would omit "(by using the Listener)" and only mention the "what" not the "how" here

File: doc/website/getting-started/examples/icedelivery.md [hunk 20]
```diff
@@ -0,0 +1,5 @@
+---
+title: How to send and receive data in a functional way
```
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply doc/website/getting-started/examples/icedelivery.md:2]: We should not mention the functional  programming paradigms here. Besides, only using "functional" here would not even make that meaning clear.

File: doc/website/getting-started/examples/icehello.md (hunk 5 above)
Comment (MatthiasKillat) [reply doc/website/getting-started/examples/icehello.md:2]: I think How to send data to another process is ok. Somehow does not sound "quite right" but is clear.
Alternatively Sending data to another process (via iceoryx, but this should be clear). Also, the overall style should be kept similar (not "How to XYZ" in some and this style in others if possible).

File: doc/website/getting-started/examples/iceoptions.md (hunk 1 above)
Comment (MatthiasKillat) [reply doc/website/getting-started/examples/iceoptions.md:2]: Here it would have to be "Configuring ..."
I do not understand what starting behavior is supposed to tell me, it is kind of confusing to a reader I would say.

File: doc/website/getting-started/examples/singleprocess.md [hunk 21]
```diff
@@ -0,0 +1,5 @@
+---
+title: How to communicate in a single process between threads
```
Comment (MatthiasKillat) [inline doc/website/getting-started/examples/singleprocess.md:2]: I do not think threads need to be mentioned here (besides, they are technically not necessary but without there might be little point except for using the same interface for every communication).

File: doc/website/getting-started/examples/waitset.md (hunk 7 above)
Comment (MatthiasKillat) [reply doc/website/getting-started/examples/waitset.md:2]: Waiting for data events

File: doc/website/getting-started/examples/waitset_in_c.md (hunk 8 above)
Comment (MatthiasKillat) [reply doc/website/getting-started/examples/waitset_in_c.md:2]: Waiting for data events in C

The waitset is the "how" again and the title covers the "what".

Or does this include user trigger events? I think this warrants its own section. The split could be similar in C++, but if we do not want the split just 

Using the WaitSet in C

would also be fine I guess.

File: doc/website/getting-started/list-of-examples.md (hunk 17 above)
Comment (MatthiasKillat) [reply doc/website/getting-started/list-of-examples.md:3]: Either lexicographically or by "complexity" which is very subjective (so probably not a good choice). The technology category column (maybe we might need a better name) would be helpful I guess.

File: doc/website/getting-started/list-of-examples.md (hunk 10 above)
Comment (MatthiasKillat) [reply doc/website/getting-started/list-of-examples.md:6]: And even this I would not mention here because it is functional programming which is meant and is misleading here.

File: doc/website/getting-started/list-of-examples.md (hunk 3 above)
Comment (MatthiasKillat) [reply doc/website/getting-started/list-of-examples.md:7]: I would call it "(Advanced) configuration options for ports"

File: doc/website/getting-started/list-of-examples.md (hunk 11 above)
Comment (MatthiasKillat) [reply doc/website/getting-started/list-of-examples.md:9]: Those match the titles in the other file I think so the comments there apply here as well.

File: doc/website/getting-started/list-of-examples.md (hunk 4 above)
Comment (MatthiasKillat) [reply doc/website/getting-started/list-of-examples.md:16]: Using the introspection to monitor the system state (or similar)

File: doc/website/getting-started/examples/iceoptions.md (hunk 1 above)
Comment (mossmaurice) [reply doc/website/getting-started/examples/iceoptions.md:2]: Done. Hope you excuse me for favouring British English :innocent: 

I'm unsure whether to use present participle here or just presence. @ithier being a native speaker, volunteered to read the documentation mid next week :100:

File: doc/website/getting-started/list-of-examples.md (hunk 3 above)
Comment (mossmaurice) [reply doc/website/getting-started/list-of-examples.md:7]: @MatthiasKillat We opted out long time ago, to call the publisher and subscribers "ports". 

I went for "Configure pub/sub settings like history cache size or startup behaviour"

File: doc/website/getting-started/examples/iceperf.md (hunk 6 above)
Comment (mossmaurice) [reply doc/website/getting-started/examples/iceperf.md:2]: @ithier Measure or Measuring?

File: doc/website/getting-started/list-of-examples.md (hunk 17 above)
Comment (mossmaurice) [reply doc/website/getting-started/list-of-examples.md:3]: Alphabetically won't help new readers, the table is structured by easy/beginner/important ... hard/advanced/unimportant

File: doc/website/getting-started/examples/waitset.md (hunk 7 above)
Comment (mossmaurice) [reply doc/website/getting-started/examples/waitset.md:2]: Active is easier to grasp than passive senteces

File: doc/website/getting-started/examples/waitset_in_c.md (hunk 8 above)
Comment (mossmaurice) [reply doc/website/getting-started/examples/waitset_in_c.md:2]: Probably the user triggers should be moved to it's own example, but not for the Almond release.

File: doc/website/getting-started/list-of-examples.md (hunk 9 above)
Comment (mossmaurice) [reply doc/website/getting-started/list-of-examples.md:5]: It's prepared on another branch. I would like to agree on the list of examples for the Almond release and then implement them.

File: doc/website/getting-started/list-of-examples.md (hunk 17 above)
Comment (mossmaurice) [reply doc/website/getting-started/list-of-examples.md:3]: With such a small number of examples, I currently don't see the benefit of a technology column. E.g. adding "WaitSet" and "Introspection" are just duplicates

File: doc/website/getting-started/examples/singleprocess.md (hunk 21 above)
Comment (mossmaurice) [reply doc/website/getting-started/examples/singleprocess.md:2]: IMHO I keep the threads in there, to put extra emphasis on them

**Review by dkroenke — APPROVED** [review]

**Review by dkroenke — APPROVED** [review]

File: CONTRIBUTING.md [hunk 22]
```diff
@@ -122,6 +122,13 @@ The folder structure boils down to:
 
 All new code should follow the folder structure.
 
+### How to add a new example
+
+1. Add the example in the ["List of all examples"](./doc/website/getting-started/list-of-examples.md)
+2. Create a new file and include the readme in `doc/website/getting-started/examples/ice-foo.md`
```
Comment (marthtz) [inline CONTRIBUTING.md:128]: I think this is not very clear. Basically you want a new readme to be added to this folder `doc/website/getting-started/examples/`, right?! It's also not clear if it always has to start with `ice`. Also, should the readme go here or will the file included here basically just point to a readme in the folder of the example?

**Review by marthtz — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md [hunk 23]
```diff
@@ -122,6 +122,13 @@ The folder structure boils down to:
 
 All new code should follow the folder structure.
 
+### How to add a new example
+
+1. Add the example in the ["List of all examples"](./doc/website/getting-started/list-of-examples.md)
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply CONTRIBUTING.md:127]: Fixed, actually I managed to include the list from `iceoryx_examples/README.md`

File: CONTRIBUTING.md (hunk 22 above)
Comment (mossmaurice) [reply CONTRIBUTING.md]: Jep, examples are included to have a single source. Clarified the text

**Review by marthtz — APPROVED** [review]

**Review by dkroenke — APPROVED** [review]
